### ISSUE TITLE
Implement verbose regular expressions

### ISSUE THREAD

**Issue by scottkleinman** [WheatonCS/Lexos#284]
Verbose regular expressions allow you to break up components of your pattern with comments to clarify what each portion of the pattern is doing. For instance, you might do something like this:

```
pattern = """
    ^                   # beginning of string
    [a-z]               # lower case ASCII alphabetic characters
    $                   # end of string
"""

re.search(pattern, 'hello', re.VERBOSE) 
```
That's not all that useful for the pattern above but could be for more complicated patterns, especially if we need to modify them later.

More information on verbose regular expressions can be found at [http://www.diveintopython.net/regular_expressions/verbose.html](http://www.diveintopython.net/regular_expressions/verbose.html).

Also [regex101.com](https://regex101.com/#python) is also very good for testing regular expressions and provides good explanations which can be pasted in as comments.

**Comment by chantisnake** [comment]
This is a super cool feature in python, I don't know that before...

It would be nice to put this in code.

**Comment by scottkleinman** [comment]
It also makes it super easy to find all the places where you are using regex.

**Comment by chantisnake** [comment]
well I don't think finding where to use regex is that big of a deal, just seach for `\ [r|ur]\".*\"\ `

**Comment by evanlaferriere** [comment]
I implemented a verbose expression in the CountVectorizer call in getMatrix

**Comment by evanlaferriere** [comment]
Correction, I thought I did so but lexos now crashes because of it. CountVectorizer takes a string as a token_pattern, and no verbose regex.

**Comment by scottkleinman** [comment]
Yes, verbose regex only works with the `re` module. That doesn't mean that what you did is useless. I recommend something like this:

```python
"""
Documentation for token_pattern:
    ur'          # For Unicode and raw patterns
    (?u)         # Use greedy matching
    \b           # Match a word boundary followed by
    [\w\'\-]+    # One or more word, apostrophe, and hyphen characters 
    \b           # Before a word boundary
"""
```
So you are basically using the comment format of verbose pattern definition to clearly indicate the meaning of each part of the pattern.  In other cases where you are actually able to use the pattern with `re`, it would look like this:

```python
token_pattern = """
    ur'          # For Unicode and raw patterns
    ...
"""
```
I think this will provide the greatest clarity for documenting regex patterns.

**Comment by mleblanc321** [comment]
small bug in evan's regex pattern;  evan has verbose regex working in countVectorizer;  we really need a sweep of all places we use regex, changing each to use a verbose mode which forces us to document

**Comment by evanlaferriere** [comment]
Moses let me know about [this](http://stackoverflow.com/questions/14885567/strip-a-verbose-python-regex), but writing functions to convert verbose to regular regex so it can passed to countVectorizer seems like overkill.

**Comment by scottkleinman** [comment]
I like the function to convert verbose to regular regex, but I also think it's probably overkill. Just providing the documentation in verbose style comments as I suggested above seems sufficient to me for cases where we are not calling `re`.

I agree that we should do a general sweep and convert to verbose regex.

**Comment by chantisnake** [comment]
There is one problem about this.

if the user do not scrub, we are facing some very strange char in the paragraph. If we don't handle them in the `count_vectorizer` that can cause error in the word count all across the whole lexos

**Comment by scottkleinman** [comment]
`CountVectorizer` is still implementing the regex:

```python
"""
Documentation for token_pattern:
    ur'          # For Unicode and raw patterns
    (?u)         # Use greedy matching
    \b           # Match a word boundary followed by
    [\w\'\-]+    # One or more word, apostrophe, and hyphen characters 
    \b           # Before a word boundary
"""

CountVector = CountVectorizer(input=u'content', encoding=u'utf-8', min_df=1,
                        analyzer=tokenType, token_pattern=ur'(?u)\b[\w\'\-]+\b',
                        ngram_range=(ngramSize, ngramSize),
                        stop_words=[], dtype=float, max_df=1.0)
```
We're just using the verbose description in comments to document the pattern in case we ever need to change it.

Or is there some other problem that I've missed?

**Comment by chantisnake** [comment]
nuh, that `ur'(?u)\b[\w\'\-]+\b'` only works when people scrub their text, if they don't scrub then the `CountVectorizer` will regard some unicode hyphen as word boundary and split that into two words.

the way to fix this is:
  1. tell everyone to scrub
  2. implement a super long regex that will be unreadable with out verbose

**Comment by scottkleinman** [comment]
Ah, I see now. I'm in the middle of a conference session, but I'll have a think about this as soon as I can.

**Comment by mleblanc321** [comment]
ur'(?u)\b[\w\'\-]+\b' only works when people scrub their text

true;   i think it is not completely unreasonable to force the workflow
i.e., tell everyone to scrub (first)
(we cannot repeat the scrubbing functionality at the point of countVectorizer)

**Comment by scottkleinman** [comment]
What if the user keeps apostrophes and/or word-internal apostrophes? Then `ur'(?u)\b[\w\'-]+\b'` will tokenise incorrectly. I think we need to test the scrubbing options and apply a different pattern depending on the setting.

We'd need a default option if the user has not set options in Scrub or Tokenize. My (pre-coffee) instinct is to use `ur'(?u)\b\w+\b'`, but to allow the user to set the default option in Settings (cog icon) if their language does not use spaces for word boundaries.

**Comment by scottkleinman** [comment]
After looking at a couple of examples, I think we may not want to use verbose syntax for short regular expressions. For instance, take `<s(.*?)>`. The verbose version might be something like this:

```python
pattern = """
    <s     # Matches the characters <s literally (case sensitive)
    .*?    # Lazily matches any character (except newline) any number of times
    >      # Matches the characters > literally
"""
```
That seems excessive. Perhaps short expressions like these should just be documented by a standard comment like:

```python
# <s(.*?)>: Lazily matches any number of characters between <s and >
```

That said, we _should_ documents all but the most basic regex patterns.

**Comment by mleblanc321** [comment]
right;  verbose is good for those killer regex;   

summary of where we are:
- we are having issues with \b (detecting word boundaries) in unicode; e.g., if a word has unicode punctuation at the end;  alvaro/jordan/kristina are all working on scrubber so we'll need to get together on this

- scrubbing XML will need some extra attention, but clearly, you've provided a good start

**Comment by scottkleinman** [comment]
I didn't know about the word boundary issue. I'm in the process of documenting the regular expressions in `processors/prepare/scrubber.py`, so the others should just follow my lead in making changes.

Stripping xml tags seems to work based on the changes I made for issue #295. So the main thing will be allowing the user to fine tune. But I think that's mainly going to be a UI task.

I think we're nearly ready to close this issue, but I want to add one more insight from the work of just done on _Dictionary of Old English_ tag stripping. For this one, I used the following pattern:

```python
pattern = re.compile(ur"""
    <        # matches the character < literally
    (.+?)    # first capturing group lazily matches any character (except 
             # newline) one or more times
    >        # matches the character > literally
    (.+?)    # first capturing group lazily matches any character (except 
             # newline) one or more times
    <        # matches the character < literally
    \/       # matches the character / literally
    \1       # matches the same text as most recently matched by the first 
             # capturing group
    >        # matches the character > literally
""", re.VERBOSE|re.MULTILINE)
text = re.sub(pattern, u'', text)
```
The point here is just to show how to use verbose syntax with `re.compile` and Unicode and raw strings.

#### So to summarise guidelines for regular expressions:

1. Short obvious patterns need no documentation.
2. Short patterns should be documented in normal python comments with the template `pattern: explanation`.
3. Long patterns should be documented and implemented using verbose syntax where possible.
4. Where a long regular expression is not used with `re`, it should still be documented in multiline comments using verbose syntax with a header like "Documentation for `pattern`:".

**Comment by mleblanc321** [comment]
we'll need an extensive set of tests to pass thru these regex;  they are deceiving simple (and easy to get wrong);
